nrf/bluetooth: Add support for nimble based ubluetooth.

andrewleech opened the pull request:
Initial support for nimble based `ubluetooth` support on nrf port. Using this replaces any need for the softdevice.

Has only been tested on nrf52840 dongle.

TODO: 
* Describing implementation choices / annoyances (eg. rtc, ticker).
* Test with default DFU on dongle - so far I've been using `make DEBUG=1 DFU=0` firmware with jlink.
* Docs updates.
* Testing on other chips / boards.
* Check performance.
* Running multitests - not sure if all features work.
* No effort has been made to support btstack yet, I believe controller support is not really available yet there?

andrewleech commented:
Now tested on nrf52840 dongle with the Adafruit UF2 bootloader: https://github.com/adafruit/Adafruit_nRF52_Bootloader/releases/download/0.6.3/pca10059_bootloader-0.6.3_s140_6.1.1.zip

Flashed that on with nrfutil following adafruit instructions
```
adafruit-nrfutil --verbose dfu serial --package pca10059_bootloader-0.6.3_s140_6.1.1.zip -p /dev/ttyACM0 -b 115200 
```

It reboots in mass storage mode.

While this bootloader comes with a copy of the softdevice bundled, micropython will happily flash over the top of that. Apparently the adafruit uf2 bootloader live checks if SD is installed and can work with/without it just fine - however no BLE OTA updates without it.

I compiled this current commit (fea9f0d0) again with DFU=1 then converted the hex to uf2
```
curl -L https://raw.githubusercontent.com/microsoft/uf2/master/utils/uf2conv.py > uf2conv.py
curl -L https://raw.githubusercontent.com/microsoft/uf2/master/utils/uf2families.json > uf2families.json
python uf2conv.py firmware.hex  -c -f 0xADA52840 -o nrf52840_micropython_ubluetooth_fea9f0d0.uf2
```
Then drag and drop the uf2 onto the board bootloader.
It copies, then reboots with micropython running.

double click the reset switch to get back into bootloader.

dpgeorge commented:
Note that there is a copy of `uf2conv.py` in this repo, under `tools/`.

andrewleech commented:
Ok, I've tried running the ble multitests... it didn't go great. Initially they all failed, found a few issues, fixed a few issues. 
Currently, running with two nrf boards, I get 6 out of 30 tests failing still.
```
micropython/tests$ python run-multitests.py -i pyb:/dev/ttyACM1 -i pyb:/dev/ttyACM0 multi_bluetooth/*
...
30 tests performed
24 tests passed
6 tests failed: multi_bluetooth/ble_gap_pair.py multi_bluetooth/ble_gap_pair_bond.py multi_bluetooth/perf_gatt_char_write.py multi_bluetooth/perf_gatt_notify.py multi_bluetooth/perf_l2cap.py multi_bluetooth/stress_log_filesystem.py
```
I haven't looked into these failures specifically yet, other than the `stress_log_filesystem` which requires `print(file=)` kwarg, which requires `MICROPY_PY_IO` which needs fixes present in my (unfinished) https://github.com/micropython/micropython/pull/8317
```

andrewleech commented:
So coming back to this, for some reason the previous approach to flashing adafruit uf2 bootloader on the nordic nrf52840 dongle won't work for me anymore. Pressing the right-angle button on the dongle puts it into bootloader mode, at which point I try:  `pip install adafruit-nrfutil; adafruit-nrfutil --verbose dfu serial --package pca10059_bootloader-0.6.3_s140_6.1.1.zip -p COM39 -b 115200 --singlebank`

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

andrewleech committed:
nrf/bluetooth: Add support for nimble based ubluetooth.

Add NimBLE bluetooth stack support for the nrf port, enabling the
ubluetooth module on nrf52 boards. This includes:

- NimBLE HCI transport and OS abstraction layer for nrf
- Soft timer and PendSV support required by the BLE stack
- HCI polling scheduled via eventq_put for efficient operation
- Board-level configuration for enabling bluetooth
- Random number generator integration for BLE security

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
py/builtinevex: Only define code_execute() when needed for build.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech commented:
It seems most multitests are unreliable, occasionally failing with a timeout.
Other tests fail pretty much every time.
So there's something not right with the integration so far.
I just tried replacing my ticker use for hci poll with pendsv and soft timer like other ports but that doesn't seem to have helped.

robert-hh commented:
> It seems most multitests are unreliable, occasionally failing with a timeout.
Other tests fail pretty much every time.

I had noticed that when adapting the NINAW10 package for the SAMD and MIMXRT port. Some tests fail always, some tests fail never, some sometimes. Using the ESP_HOSTED package is much more reliable, even if both rely on nimble. With ESP_HOSTED still some tests fail nearly always (like deepsleep and stress_log_filesystem, but the other tests pass fine. The list below shows the success rate with the NINAW10 package of 100 attempts.
```
77 ble_characteristic
100 ble_gap_advertise
100 ble_gap_connect
63 ble_gap_device_name
40 ble_gap_pair_bond
54 ble_gap_pair
55 ble_gattc_discover_services
63 ble_gatt_data_transfer
83 ble_l2cap
0 ble_mtu
57 ble_subscribe
94 perf_gatt_char_write
58 perf_gatt_notify
58 perf_l2cap
```

andrewleech commented:
That's really interesting. The tests I was having issues with seemed fine going from pyb-d to stm32wb55 so it's definitely the nrf that's got issues compared to them.

I did some more testing and suspect my soft timer integration might not be working, I'll want to get a debugger connected to check what it's actually doing.

dpgeorge commented:
For me, the BLE multitests are very robust when testing using PYBD-SFx and ESP32 boards, which use NimBLE.  Maybe occasionally one of them will fail, but it's not often.

robert-hh commented:
There are no problems in testing with PYBD or ESP32. Since both NINAW10 and ESP_HOSTED use the ESP32 implementation of BLE, but with a  different version of the ESP_IDF (3.x vs. 5.x), it looks like that make the difference.

patfrench commented:
Hello @andrewleech 
I have a custom board based on NRF52840 but in QFN48 pin version (there is no USB).

I'd like to know how to test the NRF52840_DONGLE port with NIMBLE? In my case, I don't know at which address the firmware should start.

Here is my configuration and I would like (if possible) to have the REPL over BLE. 

Can you help me? (I program with a JLINK probe)
Thanks in advance


This is my config
```
#define MICROPY_HW_BOARD_NAME “MDK-USB-DONGLE”
#define MICROPY_HW_MCU_NAME “NRF52840”
#define MICROPY_PY_SYS_PLATFORM “nrf52840-MDK-USB-Dongle”

#define MICROPY_PY_MACHINE_UART (1)
#define MICROPY_PY_MACHINE_HW_PWM (1)
#define MICROPY_PY_MACHINE_RTCOUNTER (1)
#define MICROPY_PY_MACHINE_I2C (1)
#define MICROPY_PY_MACHINE_ADC (1)
#define MICROPY_PY_MACHINE_TEMP (1)

#define MICROPY_HW_ENABLE_RNG (1)

#define MICROPY_HW_ENABLE_USBDEV (0)
#define MICROPY_HW_USB_CDC (0)

// UART config
#define MICROPY_HW_ENABLE_UART_REPL(1)
#define MICROPY_HW_UART1_RX (20)
#define MICROPY_HW_UART1_TX (19)

// SPI0 config
#define MICROPY_HW_SPI0_NAME “SPI0

#define MICROPY_HW_SPI0_SCK (10)
#define MICROPY_HW_SPI0_MOSI (7)
#define MICROPY_HW_SPI0_MISO (6)
```

andrewleech commented:
I've rebased this to current master but haven't yet re-tested anything
